### ISSUE TITLE
2 way audio support

### ISSUE THREAD

**Issue by NikDevx** [roleoroleo/yi-hack-MStar#344]
Hello @roleoroleo 
Is it possible to add 2 way audio or just audio support for this camera Yi 1080p Home 9FUS (y203c)?

**Comment by roleoroleo** [comment]
Audio is supported in the stream.
2 way is work in progress but I don't know if I will be able to complete it.

**Comment by NikDevx** [comment]
> Audio is supported in the stream.
> 2 way is work in progress but I don't know if I will be able to complete it.

Can I help you?

I saw something about 2 way audio here https://github.com/Sunoo/homebridge-camera-ffmpeg/issues/738

**Comment by roleoroleo** [comment]
The problem here is how to inject audio to the device, not how to stream it.
I'm working on a hacked libalsa but it's very complicated.

**Comment by NikDevx** [comment]
Do you mean the problem with detecting the sound card to transfer audio from the device to the camera speaker?

**Comment by roleoroleo** [comment]
Audio devices are in use by rmm process.
You can't open them with another process.
About mic, we are using an hacked version of tinyalsa that writes all samples captured to a fifo.
Now, I'm working to a similar hack to play audio on the speaker.

**Comment by NikDevx** [comment]
I got it.
Thanks

**Comment by roleoroleo** [comment]
Do you want to try a beta?

**Comment by NikDevx** [comment]
Yes
I can try tomorrow

**Comment by roleoroleo** [comment]
[y203c_0.4.1.tar.gz](https://github.com/roleoroleo/yi-hack-MStar/files/6318002/y203c_0.4.1.tar.gz)

**Comment by roleoroleo** [comment]
How it works:
1 - Create a raw pcm (wav without header): 16 KHz, 16 bit, mono (S16LE).
2 - Copy it to /tmp/audio_in_fifo

If you want to use a tts engine download nanotts from here:
https://github.com/roleoroleo/yi-hack-utils
and use with this webservice:

- http://IP-CAM:8080/cgi-bin/speak.sh?lang=en-US
- POST
- Text message in the payload

**Comment by NikDevx** [comment]
> How it works:
> 1 - Create a raw pcm (wav without header): 16 KHz, 16 bit, mono (S16LE).
> 2 - Copy it to /tmp/audio_in_fifo
> 
> If you want to use a tts engine download nanotts from here:
> https://github.com/roleoroleo/yi-hack-utils
> and use with this webservice:
> 
> * http://IP-CAM:8080/cgi-bin/speak.sh?lang=en-US
> * POST
> * Text message in the payload

Which one ver. I should use? MStar, Allwinner, Allwinner-v2

**Comment by NikDevx** [comment]
> [y203c_0.4.1.tar.gz](https://github.com/roleoroleo/yi-hack-MStar/files/6318002/y203c_0.4.1.tar.gz)

Will all my configuration data be deleted after flashing?

**Comment by roleoroleo** [comment]
Yes, save it in Maintenance page.

**Comment by NikDevx** [comment]
Hello!
This working like a charm 
Thanks a lot

**Comment by NikDevx** [comment]
Which one audio card you use for send audio to speaker?

For example I can send audio to speaker via ffmpeg
I use `alsa` audio card in my raspberry-pi camera speaker `"audio": "2way -f alsa default"`

**Maybe I can send somehow audio to speaker via ffmpeg?**

**Comment by roleoroleo** [comment]
The cam support alsa via tinyalsa lib.
But the device is busy because rmm opens it ad boot.
So, you can't use it with standard programs.
I created a "fork" patching the library: the library reads from a fifo file and send the raw audio to the speaker.

**Comment by NikDevx** [comment]
> The cam support alsa via tinyalsa lib.
> But the device is busy because rmm opens it ad boot.
> So, you can't use it with standard programs.
> I created a "fork" patching the library: the library reads from a fifo file and send the raw audio to the speaker.
> You can send audio writing to this file (/tmp/audio_in_fifo) but the stream must to match the correct format: 16 KHz, 16 bit, mono (S16LE).

I already used this one

```
http://IP-CAM:8080/cgi-bin/speak.sh?lang=en-US
POST
Text message in the payload
```

Anyway thank you!

**Comment by roleoroleo** [comment]
ffmpeg from local shell or remotely?
Where do you want to run ffmpeg?

**Comment by NikDevx** [comment]
ffmpeg on my raspberry-pi
I am using this plugin to add camera to homebridge https://github.com/Sunoo/homebridge-camera-ffmpeg#readme
And this plugin has ffmpeg

**Comment by roleoroleo** [comment]
You could try to create a new web service with this code:
```
#!/bin/sh

printf "Content-type: application/json\r\n\r\n"
printf "{}"

cat - > /tmp/audio_in_fifo
```

Pay attention to the permission 0755
Then you can send a file using this web service as a payload of a post request.
I gave it a try and it's ok .
Probably combining ffmpeg and curl it works.

**Comment by NikDevx** [comment]
> You could try to create a new web service with this code:
> 
> ```
> #!/bin/sh
> 
> printf "Content-type: application/json\r\n\r\n"
> printf "{}"
> 
> cat - > /tmp/audio_in_fifo
> ```

I should create via ssh or another way?

**Comment by roleoroleo** [comment]
https://github.com/roleoroleo/yi-hack-MStar/blob/master/src/www/httpd/cgi-bin/speaker.sh

**Comment by NikDevx** [comment]
Thanks a lot!

**Comment by NikDevx** [comment]
Can you compile firmware with add speaker service please?

**Comment by roleoroleo** [comment]
Check the release section.

**Comment by NikDevx** [comment]
> Check the release section.

You are the best!
Thanks!

**Comment by bepece1** [comment]
How can I use the new "speaker" function in version 0.4.1???

**Comment by roleoroleo** [comment]
**speak.sh**
http://IP-CAM:8080/cgi-bin/speak.sh?lang=en-US
POST
Text message in the payload

**speaker.sh**
http://IP-CAM:8080/cgi-bin/speaker.sh
POST
Raw audio in the payload: 16 KHz, 16 bit, mono (S16LE)

**Comment by jd1900** [comment]
Any idea of how to tie this with TTS from Home Assistant to send audio to the camera?

**Comment by NikDevx** [comment]
> **speak.sh**
> http://IP-CAM:8080/cgi-bin/speak.sh?lang=en-US
> POST
> Text message in the payload
> 
> **speaker.sh**
> http://IP-CAM:8080/cgi-bin/speaker.sh
> POST
> Raw audio in the payload: 16 KHz, 16 bit, mono (S16LE)

@roleoroleo, on actual version of firmware speak and speaker not working.

my curl :`curl -X POST -d 'hello' http://192.168.0.30:8080/cgi-bin/speak.sh?lang=en-US`
`zsh: no matches found: http://192.168.0.30:8080/cgi-bin/speak.sh?lang=en-US`

work this `curl -X POST -d 'hello' http://192.168.0.30:8080/cgi-bin/speak.sh`
```
{
"error":"false",
"description":"hello"
}
```
But no audio coming

**Comment by NikDevx** [comment]
> Any idea of how to tie this with TTS from Home Assistant to send audio to the camera?

This issue not about home assistant
Open new issue and ask

**Comment by roleoroleo** [comment]
Install this integration:
https://github.com/roleoroleo/yi-hack_ha_integration/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc
and use speak service.

**Comment by roleoroleo** [comment]
> 
> 
> Any idea of how to tie this with TTS from Home Assistant to send audio to the camera?

Install this integration:
https://github.com/roleoroleo/yi-hack_ha_integration/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc
and use speak service.

**Comment by NikDevx** [comment]
> > **speak.sh**
> > http://IP-CAM:8080/cgi-bin/speak.sh?lang=en-US
> > POST
> > Text message in the payload
> > **speaker.sh**
> > http://IP-CAM:8080/cgi-bin/speaker.sh
> > POST
> > Raw audio in the payload: 16 KHz, 16 bit, mono (S16LE)
> 
> @roleoroleo, on actual version of firmware service speak and speaker not working.
> 
> my curl: `curl -X POST -d 'hello' http://192.168.0.30:8080/cgi-bin/speak.sh?lang=en-US`
> `zsh: no matches found: http://192.168.0.30:8080/cgi-bin/speak.sh?lang=en-US`
> 
> only work without `lang` parameter `curl -X POST -d 'hello' http://192.168.0.30:8080/cgi-bin/speak.sh`
> 
> ```
> {
> "error":"false",
> "description":"hello"
> }
> ```
> 
> But no audio coming
> 
> ↓↓↓↓↓
> 
> > [y203c_0.4.1.tar.gz](https://github.com/roleoroleo/yi-hack-MStar/files/6318002/y203c_0.4.1.tar.gz)
> 
> Here it was working

@roleoroleo, any updates?

**Comment by roleoroleo** [comment]
The only change is this commit: https://github.com/roleoroleo/yi-hack-MStar/commit/3762a570ec91777162228b93d2ac7d4b26e0c1f1
but it's not related.
And with my y203c it works.
I reflashed the 0.4.1 now and tested it with curl.

**Comment by roleoroleo** [comment]
Ah...
I moved the path of the tts libs.
/tmp/sd/yi-hack/usr/share...

Before it was:
/tmp/sd/yi-hack/share...

**Comment by NikDevx** [comment]
> Ah...
> I moved the path of the tts libs.
> /tmp/sd/yi-hack/usr/share...
> 
> Before it was:
> /tmp/sd/yi-hack/share...

Thanks 🥇

**Comment by NikDevx** [comment]
Okay
Service speak working now!

But speaker not play my wav file.

`curl -X POST -F "file=http://192.168.0.30:8080/record/open.wav" http://192.168.0.30:8080/cgi-bin/speaker.sh`
I get in response `{}`
And I hear some crackle

**Comment by roleoroleo** [comment]
You need to remove the header from the wav file.
And check if the format is correct.

**Comment by NikDevx** [comment]
> You need to remove the header from the wav file.
> And check if the format is correct.

How I can remove header?

Can you give me you wav file for testing speaker, please.

**Comment by roleoroleo** [comment]
Remove first 44 bytes.

**Comment by roleoroleo** [comment]
You can export a raw data with Audacity.

16 KHz, 16 bit, mono (S16LE)

**Comment by NikDevx** [comment]
> Okay
> Service speak working now!
> 
> But speaker not play my wav file.
> 
> `curl -X POST -F "file=http://192.168.0.30:8080/record/open.wav" http://192.168.0.30:8080/cgi-bin/speaker.sh`
> I get in response `{}`
> And I hear some crackle


My open.raw
16 KHz, 16 bit, mono (S16LE)

The same
`curl -X POST -F "file=http://192.168.0.30:8080/record/open.raw" http://192.168.0.30:8080/cgi-bin/speaker.sh`
I get in response {}
And I hear some crackle


The same raw file in `/tmp/audio_in_fifo` work perfect
`curl -X POST -d 'hello' http://192.168.0.30:8080/cgi-bin/speak.sh`
```
{
"error":"false",
"description":"hello"
}
```

**Comment by NikDevx** [comment]
I want to make sure I understand correctly
I can send an audio file to the speaker service and it will play my audio file?

**Comment by foxel** [comment]
@roleoroleo thanks for the thing. Works great.

The only two things I want to say:
* Please put the instructions for TTS setup somewhere where it's simpler to find. E.g. having your HA integration gives access to TTS immediately even if TTS engine is hot installed
* Is the language set limited by the library? Are there other options for TTS engine which include russian?

**Comment by foxel** [comment]
@NikDevx the issue is in your POST payload.
try the following assuming you have `open.raw` file in working directory:
```shell
curl -X POST --data-binary "@open.raw" "http://192.168.0.30:8080/cgi-bin/speaker.sh"
```

**Comment by foxel** [comment]
BTW. I was able to use HASS Cloud TTS with YI-hack [here](https://github.com/foxel/yi-hack_ha_integration/tree/hass_cloud_tts).
@roleoroleo this also has some fixes in terms of undefined variables in service calls

**Comment by roleoroleo** [comment]
>     * Please put the instructions for TTS setup somewhere where it's simpler to find. E.g. having your HA integration gives access to TTS immediately even if TTS engine is hot installed
> 
>     * Is the language set limited by the library? Are there other options for TTS engine which include russian?

I will add a wiki page.

Languages are limited by the picotts/nanotts library: en, de, es fr and it. But you could find additional packages.
Check this page: https://remzhuk.ru/en/russkii-yazyk-dlya-pico-tts-obzor-sintezatorov-rechi-dlya-android-sintez/

**Comment by roleoroleo** [comment]
> 
> 
> BTW. I was able to use HASS Cloud TTS with YI-hack [here](https://github.com/foxel/yi-hack_ha_integration/blob/9003d6c0a2ad3943bd3a782569cfe036e0940347/custom_components/yi_hack/camera.py#L362-L404).
> ![image](https://user-images.githubusercontent.com/1551926/116350077-842c2100-a81b-11eb-8eef-ebb69a97715c.png)
> 
> @roleoroleo this also has some fixes in terms of undefined variables in service calls

Thanks for your work, I will check it and add it to my repo, if you agree.

**Comment by foxel** [comment]
> Thanks for your work, I will check it and add it to my repo, if you agree.

No problem. You can merge that code. The only reasons stopping me from having a PR right now are the following:
* my solution replaces yours with built-in TTS (maybe it's better to have it optional)
* it requires HASS Cloud subscription
* getting PCM audio with needed parameters is a bit of a hack around hass_nabucasa code which is hardcoded to produce mp3

**Comment by foxel** [comment]
Another observation: CGI script playing the audio (`speaker.sh`) streams the body right into the fifo. This has the following effects:
* audio hicckups if data transmission has some lags (encountered only once)
* HTTP request is not closed until the audio is played.

As a fix I use slightly modified `speaker.sh`:
```shell
#!/bin/sh

printf "Content-type: application/json\r\n\r\n"
printf "{}"

cat - > /tmp/speaker_temp

cat /tmp/speaker_temp > /tmp/audio_in_fifo &
```

Not ideal, but works.

**Comment by roleoroleo** [comment]
> Another observation: CGI script playing the audio (`speaker.sh`) streams the body right into the fifo. This has the following effects:
> 
>     * audio hicckups if data transmission has some lags (encountered only once)
> 
>     * HTTP request is not closed until the audio is played.
> 
> 
> As a fix I use slightly modified `speaker.sh`:
> 
> ```shell
> #!/bin/sh
> 
> printf "Content-type: application/json\r\n\r\n"
> printf "{}"
> 
> cat - > /tmp/speaker_temp
> 
> cat /tmp/speaker_temp > /tmp/audio_in_fifo &
> ```
> 
> Not ideal, but works.

The only problem with this solution is the memory.
Probably we should set a limit on the size of the file.
Otherwise the cam may crash and reboot.

**Comment by roleoroleo** [comment]
Proposal for a different approach.

We could try to use the standard tts service embedded in ha.
If I'm not wrong the service returns a json like this:
```
{
    "path": "/api/tts_proxy/265944c108cbb00b2a621be5930513e03a0bb2cd_en_-_demo.mp3",
    "url": "http://127.0.0.1:8123/api/tts_proxy/265944c108cbb00b2a621be5930513e03a0bb2cd_en_-_demo.mp3"
}
```
(https://www.home-assistant.io/integrations/tts/)

The component should get mp3 and convert it to the supported format.
Then, send the last file to the web service of the cam.

We should create a new "media_player" platform file with play_media capability.

**Comment by foxel** [comment]
Yaeh, that's an option too. Was thinking about it also. I feel something wrong with converting things back and forth, though,

**Comment by NikDevx** [comment]
> @NikDevx the issue is in your POST payload.
> try the following assuming you have `open.raw` file in working directory:
> 
> ```shell
> curl -X POST --data-binary "@open.raw" "http://192.168.0.30:8080/cgi-bin/speaker.sh"
> ```

> @NikDevx the issue is in your POST payload.
> try the following assuming you have `open.raw` file in working directory:
> 
> ```shell
> curl -X POST --data-binary "@open.raw" "http://192.168.0.30:8080/cgi-bin/speaker.sh"
> ```

Thank you!
I had an error when trying to open a file from an SD card.
When I open the file from raspberry-pi it works now.

**Comment by roleoroleo** [comment]
@foxel 
I'm working on media_player platform.
Do you think it's possible to converto mp3 to raw pcm using ffmpeg (inside ha)?
Or maybe is it possible to install pydub?

**Comment by roleoroleo** [comment]
@foxel 
What do you think about this alpha?
The conversion part is very ugly...

```
"""Support for output tts to the yi-hack cam."""
import asyncio
import logging
import subprocess

import requests
from requests.auth import HTTPBasicAuth
from tempfile import NamedTemporaryFile
import os

from homeassistant.components.ffmpeg import CONF_EXTRA_ARGUMENTS, DATA_FFMPEG
from homeassistant.components.media_player import DEVICE_CLASS_SPEAKER, MediaPlayerEntity
from homeassistant.components.media_player.const import (
    SUPPORT_PLAY_MEDIA,
    MEDIA_TYPE_MUSIC,
)
from homeassistant.helpers.device_registry import CONNECTION_NETWORK_MAC
from homeassistant.const import (
    CONF_HOST,
    CONF_MAC,
    CONF_NAME,
    CONF_PASSWORD,
    CONF_PORT,
    CONF_TOKEN,
    CONF_USERNAME,
)

from .const import (
    CONF_SERIAL,
    DEFAULT_BRAND,
    DOMAIN,
)


SUPPORT_YIHACK_MEDIA = (
    SUPPORT_PLAY_MEDIA
)

_LOGGER = logging.getLogger(__name__)

async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up the Samsung TV from a config entry."""

    async_add_entities([YiHackMediaPlayer(config_entry)])


class YiHackMediaPlayer(MediaPlayerEntity):
    """Define an implementation of a Yi Camera media player."""

    def __init__(self, config):
        """Initialize the device."""
        self._device_name = config.data[CONF_NAME]
        self._name = self._device_name + "_media_player"
        self._unique_id = self._device_name + "_mpca"
        self._mac = config.data[CONF_MAC]
        self._serial_number = config.data[CONF_SERIAL]
        self._is_on = True
        self._host = config.data[CONF_HOST]
        self._port = config.data[CONF_PORT]
        self._user = config.data[CONF_USERNAME]
        self._password = config.data[CONF_PASSWORD]
        # Assume that the media player is not muted
        self._muted = False
        # Assume that the media player is in Play mode
        self._playing = True
        self._state = None

    @property
    def brand(self):
        """Camera brand."""
        return DEFAULT_BRAND

    @property
    def name(self):
        """Return the name of the camera."""
        return self._name

    @property
    def unique_id(self) -> str:
        """Return the unique ID of the camera."""
        return self._unique_id

    @property
    def state(self):
        """Return the state of the camera."""
        return self._state

    @property
    def device_info(self):
        """Return device specific attributes."""
        return {
            "name": self._device_name,
            "connections": {(CONNECTION_NETWORK_MAC, self._mac)},
            "identifiers": {(DOMAIN, self._serial_number)},
            "manufacturer": DEFAULT_BRAND,
            "model": DOMAIN,
        }

    @property
    def is_volume_muted(self):
        """Boolean if volume is currently muted."""
        return False

    @property
    def supported_features(self):
        """Flag media player features that are supported."""
        return SUPPORT_YIHACK_MEDIA

    @property
    def device_class(self):
        """Set the device class to SPEAKER."""
        return DEVICE_CLASS_SPEAKER

    async def async_play_media(self, media_type, media_id, **kwargs):
        """Send the play_media command to the media player."""

        def _fetch_file():
            """Get mp3 file."""

            try:
                response = requests.get(media_id, timeout=5, allow_redirects=True)
                if response.status_code < 300:
                    return response
            except requests.exceptions.RequestException as error:
                _LOGGER.error(
                    "Error getting mp3 file: %s",
                    error,
                )

            return None

        def _perform_speaker(file_name):
            auth = None
            if self._user or self._password:
                auth = HTTPBasicAuth(self._user, self._password)

            with open(file_name, 'rb') as f:
                data = f.read()

            try:
                response = requests.post("http://" + self._host + ":" + self._port + "/cgi-bin/speaker.sh", data=data, timeout=10, headers={'Content-Type': 'application/octet-stream'}, auth=auth)
                if response.status_code >= 300:
                    _LOGGER.error("Failed to send speaker command to device %s", self._host)
            except requests.exceptions.RequestException as error:
                _LOGGER.error("Failed to send speaker command to device %s: error %s", self._host, error)

            if response is None:
                _LOGGER.error("Failed to send speak command to device %s: error unknown", self._host)

        if media_type != MEDIA_TYPE_MUSIC:
            _LOGGER.error(
                "Invalid media type %s. Only %s is supported",
                media_type,
                MEDIA_TYPE_MUSIC,
            )
            return

        _LOGGER.error("Inizio")
        in_file = NamedTemporaryFile(suffix=".mp3", delete=False)
        in_file.close()
        in_file_path = in_file.name

        r = await self.hass.async_add_executor_job(_fetch_file)
        open(in_file_path, 'wb').write(r.content)
        _LOGGER.error("Preso file " + in_file_path)

        out_file = NamedTemporaryFile(suffix=".raw", delete=False)
        out_file.close()
        out_file_path = out_file.name
        _LOGGER.error("Creato file 2 " + out_file_path)

        cmd = ["ffmpeg",  "-i",  in_file_path, "-f", "s16le", "-acodec",  "pcm_s16le", "-ar", "16000", "-y", out_file_path]
        subprocess.call(cmd, shell=False)
        _LOGGER.error("Eseguito ffmpeg")

        await self.hass.async_add_executor_job(_perform_speaker, out_file_path)
        _LOGGER.error("Eseguito ws")

        try:
            os.remove(in_file_path)
        except OSError as error:
            _LOGGER.error("Error deleting temp file %s", in_file_path)

        try:
            os.remove(out_file_path)
        except OSError as error:
            _LOGGER.error("Error deleting temp file %s", out_file_path)
```

**Comment by foxel** [comment]
@roleoroleo looks like it should work, will test tomorrow.
Btw. Looks like you can do this without temp files with ffmpeg http protocol
https://www.ffmpeg.org/ffmpeg-protocols.html#http

**Comment by roleoroleo** [comment]
I created a new "media_player" branch.

**Comment by foxel** [comment]
@roleoroleo you forgot to push the code to this new branch )

**Comment by foxel** [comment]
I've tested the code. Works. But subprocess call is sncronous and needs to be offloaded to o\another thred like so:

```
        def perform_cmd():
            subprocess.call(cmd, shell=False)

        await self.hass.async_add_executor_job(perform_cmd)
```

My experiment with HTTP protocol inside ffmpeg was not successful. I can read but output fails for some reason... will investigate some more... My most successful call is:
```
ffmpeg -i http://192.168.0.18:8123/api/tts_proxy/d4d4fe489e9464f8e400a5baa9fc28ccbe1fd472_en_-_google_translate.mp3 -f s16le -acodec pcm_s16le -ar 16000 -chunked_post 0 -method POST -content_type application/octet-stream http://192.168.0.217:8080/cgi-bin/speaker.sh -loglevel debug
```

**Comment by roleoroleo** [comment]
> @roleoroleo you forgot to push the code to this new branch )

You are right... sorry.

**Comment by roleoroleo** [comment]
I found an intermediate solution.
ffmpeg reads directly form http url but the output is captured into a local var.
It works without temp files.

Pushed...

**Comment by vin-w** [comment]
@roleoroleo 
Just tested, and it works well with HASS google_tts.

But the sound for first few sec was a little bit crack.
e.g. First part of " I am sleepy" is not as clear as second part.

Another thought, we could also add an option for convert volume loudness with ffmpeg "-vol" and add an option in the integration too (I am not sure we can control the voice level in the cam)

```
Text-to-Speech (TTS): Say an TTS message with google_translate
Message: I am sleepy, I am sleepy
lang: en
```

**Comment by roleoroleo** [comment]
AFAIK the volume is not available.
We should increase the amplitude in the pcm.

**Comment by vin-w** [comment]
<img width="841" alt="Screenshot 2021-05-01 at 2 43 16 AM" src="https://user-images.githubusercontent.com/10348873/116740163-3552d700-aa27-11eb-8426-59f3f0228ab6.png">

I am now pasting the mp3 url and let the media_player to do conversion
For the option, I think would be good to add "-ac 1" to restrict the output to mono.

```
        cmd = ["ffmpeg",  "-i",  media_id, "-f", "s16le", "-acodec",  "pcm_s16le", "-ac", "1",  "-ar", "16000", "-"]
```

**Comment by roleoroleo** [comment]
About crackles, check this solution:
https://github.com/roleoroleo/yi-hack-MStar/issues/344#issuecomment-828174129

**Comment by vin-w** [comment]
> About crackles, check this solution:
> [#344 (comment)](https://github.com/roleoroleo/yi-hack-MStar/issues/344#issuecomment-828174129)

I just found out that reboot the cam will actually solve the problem.
No need the script changes.

**Comment by roleoroleo** [comment]
Check the last code of speaker.sh: https://github.com/roleoroleo/yi-hack-MStar/blob/master/src/www/httpd/cgi-bin/speaker.sh
If sd exists, it creates a temp file in sd.
Otherwise, it check for size and it creates a temp file in /tmp

**Comment by tunerooster** [comment]
I just found this, and I am needing it...  Ie, to refer a tts command to my camera via a media_player entity.

Maybe I am missing it, but I can't see/find where/how to install this feature, and then how to access it from within HA config files.  Can you point me in the right direction?

Thanks much...

**Comment by roleoroleo** [comment]
About firmware, you only need to install the last version (0.4.3).
To connect this feature with ha, I created a new integration: https://github.com/roleoroleo/yi-hack_ha_integration
Take a look at the readme.

**Comment by tunerooster** [comment]
l have speak working fine.  It's great!  It is the media_player part I am missing.  Does the media_player interface also work on yi-hack-Allwinner-v2 (0.1.9)?  If so, can you point me there instead?  Thanks, you're the best!

**Comment by tunerooster** [comment]
Is all I need is the "yi-hack Home Assistant integration"?  I see that it is available for yi-hack-Allwinner-v2.  Is this what makes the camera available to HA as a media_player?  I don't see any mention of media_player in the README.

**Comment by tunerooster** [comment]
Ah, I see the yi-hack_ha_integration appears to make the camera available as a media_player!  Just what I need.  I will install it now and report back...

**Comment by tunerooster** [comment]
Hmm...  I'm having trouble installing "yi-hack Home Assistant integration" (per the install instructions):
- I have successfully installed the files in the custom_components/yi_hack directory
- All files show as listed in the instructions (readme)
- Restarted HA
- In _Configuration -> Integrations_, clicked Add Integration
- Then the _Set up new integration_ window pops up
- I enter "yi_hack" in the _Search Integrations_ input field
- **_Nothing_** shows up

Also, browsing the available integrations, I do not see the yi_hack integration.

I looked for a HACS repository, but found nothing there.

Please advise...

**Comment by tunerooster** [comment]
UPDATE.  Progress...

I restarted the HA server, but did not restart the browser instance.  After restarting the browser instance, it successfully finds yi_hack integration.  You may wish to mention to restart both server and browser in the readme.  Or maybe one only needs to restart the browser...?

**Comment by tunerooster** [comment]
SUCCESS  It works!  Thanks again!

**Comment by tunerooster** [comment]
In testing, I found the volume level of the camera audio output via tts to be a little weak.  I see there was a mention of upping the volume through ffmpeg.  Did that ever get implemented, and if so, can you advise how to do it?
If not, are there still plans to implement it?

I have found that sending pcm files directly to the audio fifo to be loud and clear, and I *am* able to adjust their volume with ffmpeg.  So I think the ffmpeg proposal would be effective.

**Comment by tunerooster** [comment]
I hate to say it, but the volume through the media_player interface is too low to be usable.  Using both tts and play_media the volume is to weak.  I can likely get around  play_media by increasing the volume in the .mp3 file, but I can't see a way around this using tts.  Any advice is welcome...

Thanks!

**Comment by tunerooster** [comment]
I seem to remember seeing " -rtsp_transport tcp" when installing the integration.  Could I add -filter:a "volume=25dB" there?  I can't find where/how to "reinstall" or "reconfigure" the integration.  Do I have to delete it and reinstall to get back to the field which had "-rtsp_transport tcp" in it?

Sorry for the multiple posts...

**Comment by tunerooster** [comment]
I found the "extra_arguments" option in:
```
.storage/core.config_entries
```
Should I stop HA, change it there and restart?

```
{
    "entry_id": "c1ec7feea03a21153635f45a56654e02",
    "version": 1,
    "domain": "yi_hack",
    "title": "yi_hack_a2_3aed03",
    "data": {
        "host": "frontdoor.krautclan.com",
        "port": 8080,
        "username": "root",
        "password": "xxxxxxxxx",
        "extra_arguments": "-rtsp_transport tcp",
        "SERIAL_NUMBER": "00000000000000000000",
        "mac": "xx:xx:xx:xx:xx:xx",
        "PTZ": "no",
        "HACK_NAME": "yi-hack-allwinner-v2",
        "name": "yi_hack_a2_3aed03",
        "RTSP_PORT": "554",
        "MQTT_PREFIX": "frontdoor",
        "TOPIC_BIRTH_WILL": "status",
        "TOPIC_MOTION": "motion",
        "TOPIC_SOUND_DETECTION": "sound_detection",
        "TOPIC_BABY_CRYING": "baby_crying",
        "TOPIC_MOTION_IMAGE": "",
        "MOTION_START_MSG": "motion_start",
        "MOTION_STOP_MSG": "motion_stop",
        "BIRTH_MSG": "online",
        "WILL_MSG": "offline",
        "BABY_CRYING_MSG": "crying",
        "SOUND_DETECTION_MSG": "sound_detected"
    },
    "options": {},
    "system_options": {
        "disable_new_entities": false
    },
    "source": "user",
    "connection_class": "local_poll",
    "unique_id": "xx:xx:xx:xx:xx:xx",
    "disabled_by": null
}
```

**Comment by tunerooster** [comment]
Still working on this...

I ran a test and using "ps" I was able to see where the yi-hack Home Assistant integration runs ffmpeg as follows:
```
ffmpeg -i https://mydomain.com/api/tts_proxy/a94a8fe5ccb19ba61c4c0873d391e987982fbbd3_en_-_google_translate.mp3 -f s16le -acodec pcm_s16le -ar 16000 -
```
However, I do not see the "-rtsp_transport tcp" option in the call to ffmpeg, though it was specified when the integration was initially configured as confirmed in the _.storage/core.config_entries_ file above.

What am I missing here?

Is "extra_arguments" really being used/honored or is this a bug, or am I completely off base?

Thanks for any help!

**Comment by tunerooster** [comment]
FYI, I just noticed, in the alpha python code above, the line:
```
 cmd = ["ffmpeg",  "-i",  in_file_path, "-f", "s16le", "-acodec",  "pcm_s16le", "-ar", "16000", "-y", out_file_path]
```
make no reference to the extra_arguments, and shows only the options I see in the "ps" above.

**Comment by roleoroleo** [comment]
You are right.
I probably forgot to call the variable.
I'll check it.

**Comment by tunerooster** [comment]
Thanks!  There's also another issue with the camera media player not
showing up as a Chromecast device.  Ie, I would like to be able to say "Hey
Google, Broadcast to frontdoor, Hi, I'll be right there." but the media
player camera "frontdoor" is not recognized.  The command "catt scan" dies
not list it either.  ...don't know if this is fixable, but it would be a
nice feature.  Should I open a new issue for this?

**Comment by tunerooster** [comment]
P.S., I noticed that the Google Home media_players are listed with "Google Cast" integration, whereas the media_player.frontdoor_camera is listed with "Yi Cam with yi-hack" integration.

**Comment by roleoroleo** [comment]
> It looks like to work, the camera has to be "discoverable" as a chromecast device.

I think this isn't a simple task.
But I will study it.

**Comment by roleoroleo** [comment]
Now I remember.
The extra_argument parameter is useful for video capture process.
I think it's not correct to mix these options in the ffmpeg call.
But I added a new parameter "Boost speaker" that you can check during integration (default true).
Do you want to try the last commit?

**Comment by tunerooster** [comment]
Certainly.  Not sure I know how to do that though...

Is the boost parameter an on/off switch or a value?  I found I had to set:
"-filter:a volume=3" to be effective.

Is there a release file somewhere?

**Comment by vin-w** [comment]
I was using
```
        cmd = ["ffmpeg",  "-i",  media_id, "-f", "s16le", "-acodec",  "pcm_s16le", "-ac", "1",  "-vol", "1024", "-ar", "16000", "-"]
```
```
-ac channels        set number of audio channels
-ac 1 means change to mono 

-vol volume change audio volume (256=normal)
-vol 1024 means... 1024/256 = 4 times larger
```

I only change for mstar camera, for allwinner camera the volume was loud enough

Could we make boost audio an extra config. like video capture extra_argument?

**Comment by roleoroleo** [comment]
> Is there a release file somewhere?

It's not a release but you can download the repo.

> Could we make boost audio an extra config. like video capture extra_argument?

Yes, it's an option:
![immagine](https://user-images.githubusercontent.com/39277388/142370843-0101a1be-4f7b-4f81-bcfe-66bb61c9139a.png)

**Comment by roleoroleo** [comment]
But I need help to add the right parameter for all models: mstar, allwinner and allwinner-v2.

**Comment by tunerooster** [comment]
As I mentioned above, for Allwinner-v2, I had to use:
```
ffmpeg -filter:a volume=3
```
The "3" means 3x the original volume.

However, in my opinion, it would be nice to allow a text field to let the user specify any options needed.  The "Boost" check box could be in addition, if the user does not know what ffmpeg arguments to use.

**Comment by tunerooster** [comment]
Is the "repo" you mention also available for allwinner-v2 at:
```
https://github.com/roleoroleo/yi-hack-Allwinner-v2
```
Do I just select "Download ZIP" under the Code menu?

**Comment by tunerooster** [comment]
Not sure this is your problem, but do you know of a way to reconfigure the yi-hack_ha_integration or do I need to delete it and reinstall each time?

**Comment by tunerooster** [comment]
I think I answered my own questions:
- Yes, you install it once for each camera.
- By installing it for camera 2, it the new code applies to all existing cameras.
- Didn't try, but it appears the a fresh install is the same as a reinstall.  (If I am wrong on that, someone please correct me.)
- No need to delete, just install new code under new custom_components/yi_hack directory.

**Comment by roleoroleo** [comment]
> And, if extra_arguments is not used, is should not appear. Is there a reason it is still in the "yi-hack configuration" window shown above?

extra_arguments is used to take the snapshot and to get stream. I can't remove it.

> Not sure this is your problem, but do you know of a way to reconfigure the yi-hack_ha_integration or do I need to delete it and reinstall each time?

No, I think there is no way.

> And with multiple cameras, do you install it over again, once for each camera?

Yes, once for each camera.

When you update, you don't need to remove and reinstall the cams.
The new "boost" option is default true for all cams installed before the upgrade.

**Comment by roleoroleo** [comment]
So...
Mstar:
`-ac 1 --filter:a volume=4`
Allwinner:
nothing
Allwinner-v2:
`-filter:a volume=3`

vol option seems to be deprecated.

**Comment by tunerooster** [comment]
"changing the boolean option with a integer multiplier" would be great!  That's all that is needed.  Please post when this is implemented and I will download and test it.  I downloaded the current code and it works great.  Looking forward to the release which lets me set volume=[1234] etc!

Thanks you have really created a great hack.  No other camera hack even comes close!

**Comment by tunerooster** [comment]
Update:

The media_player entity from the "yi-hack_ha_integration" seems to have quit working since I updated.  Entering "test" in the tts screen (see attached) does nothing.  I have confirmed the camera and audio are on (ie, it plays a pcm through the fifo fine), but tts plays nothing now.  Hope this isn't something I have done wrong.

I do see ffmpeg running when I click "send" as follows:
```
ffmpeg -i https://mydomain.com/api/tts_proxy/a94a8fe5c...987982fbbd3_en_-_google_translate.mp3 -f s16le -acodec pcm_s16le -ar 16000 -filter:a "volume=25dB" -
```
but no sound results from the camera.

Can you suggest any other troubleshooting tests?
[output.pdf](https://github.com/roleoroleo/yi-hack-MStar/files/7569475/output.pdf)

**Comment by roleoroleo** [comment]
Do you want to try the last commit?

**Comment by tunerooster** [comment]
Yes, will do...

**Comment by tunerooster** [comment]
Installed last commit,,,

tts.google_say is now working!

However, media_player.play_media is not.  From the Developer Tools -> SERVICES menu:
```
service: media_player.play_media
data:
  media_content_id: /local/audio/hello.mp3
  media_content_type: music
target:
  entity_id:
    - media_player.backdoor_camera
```
Produces no sound.   Whereas it works when entity_id is a Google Home Mini.

The volume via TTS is now loud and clear (at 3x) and I can see it is correctly set when ffmpeg runs.

Please advise re: media_player.play_media

Thanks as always!

**Comment by tunerooster** [comment]
Hold on...

I just discovered it is working on the frontdoor camera but not the backdoor camera.  They are supposed to be identically configured and are the same model.

I deleted and reinstalled the integration for the backdoor camera, but not for the frontdoor.  I am not clear where/what changes in a reinstall.  I.e., they both are using the same new media_player.py.  Should I delete and reinstall the frontdoor camera and see if it breaks it?

**Comment by tunerooster** [comment]
I also noticed this, from `.storage/core.config_entries`:

```
                    "host": "backdoor",
                    "boost_speaker": "x 3",
----------------
                    "host": "frontdoor",
```
I.e., no "boost_speaker" is included in the frontdoor config.

I doubt this is relevant, as it likely only applies to the tts service, but it is the only difference I could find in the configurations.   I will reinstall the frontdoor with the x3 boost and see if it causes the frontdoor to quit working via media_player.play_media service.

**Comment by tunerooster** [comment]
OK, I reinstalled the frontdoor with the x3 boost and it is still working, both play_media and tts.  So there must be something misconfigured on my backdoor camera.  I will keep looking...

Sorry for the premature alarm.

Any ideas/suggestions which may come to mind are welcome.  :)

**Comment by tunerooster** [comment]
I think I have found the problem...

For the backdoor, in the media_content_id: field, in was using "/local/audio/hello.mp3".

For the frontdoor, I was using "https://mydomain.com/local/audio/hello.mp3"

When I switch to the Fully Qualified Domain name for the backdoor, it worked.

I was certain that HA appended the domain name (base_url from the configuration.yaml file) when the media_content_id started with a slash (/), but apparently not.

Live and learn...  and sorry for the false alarm!

**Comment by roleoroleo** [comment]
I think /local/audio (without http) refers to the local file system.
If I donwload the mp3 file to /tmp and I play it with
media_content_id: /tmp/hello,mp3
it works.
Probably it depends on how media_id is handled.
In my case is sent directly to ffmpeg.

**Comment by tunerooster** [comment]
But it works one way with google home speakers and another with the camera
speaker.   It should be  the same.
